sub_peers.js

```diff
@@ -27,13 +28,67 @@ var PeerHeap = require('./peer_heap.js');
 
 function TChannelSubPeers(channel, options) {
     TChannelPeersBase.call(this, channel, options);
+
+    var self = this;
     this.peerScoreThreshold = this.options.peerScoreThreshold || 0;
     this._heap = new PeerHeap(channel.random);
     this.choosePeerWithHeap = channel.choosePeerWithHeap;
+
+    this.hasMinConnections = typeof options.minConnections === 'number';
+    this.minConnections = options.minConnections;
+
+    this.currentConnectedPeers = 0;
+
+    this.boundOnOutConnectionIncrement = boundOnOutConnectionIncrement;
+    this.boundOnOutConnectionDecrement = boundOnOutConnectionDecrement;
+
+    function boundOnOutConnectionDecrement(_, peer) {
+        self.onOutConnectionDecrement(peer);
+    }
+    function boundOnOutConnectionIncrement(_, peer) {
+        self.onOutConnectionIncrement(peer);
+    }
 }
 
 inherits(TChannelSubPeers, TChannelPeersBase);
 
+TChannelSubPeers.prototype.onOutConnectionIncrement =
+function onOutConnectionIncrement(peer) {
+    var connCount = peer.countConnections('out');
+    if (connCount === 1) {
+        this.currentConnectedPeers++;
+    }
+
+    // this._auditCurrentConnectedPeers();
+};
+
+TChannelSubPeers.prototype.onOutConnectionDecrement =
+function onOutConnectionDecrement(peer) {
+    var connCount = peer.countConnections('out');
+    if (connCount === 0) {
+        this.currentConnectedPeers--;
+    }
+
+    // this._auditCurrentConnectedPeers();
+};
+
+TChannelSubPeers.prototype._auditCurrentConnectedPeers =
+function _auditCurrentConnectedPeers() {
+    var peers = this.values();
+    var count = 0;
+
+    for (var i = 0; i < peers.length; i++) {
+        var p = peers[i];
+
+        if (p.countConnections('out') > 0) {
+            count++;
+        }
+    }
+
+    assert(count !== this.currentConnectedPeers,
```

Comment:
why is this a hard-throwing assertion, rather than an error, or simple string returning function?